tool: add tool to deal with vagrant deployments, drop 'setup.sh'

jecluis opened the pull request:
Inspired by sesdev.

Hopefully will consolidate things a bit more for us, and especially newcomers, than a bunch of scripts spread throughout.

Stretch goal, eventually, will be to also get image builds going through this script.

Signed-off-by: Joao Eduardo Luis \<joao@suse.com>

l-mb reviewed (approved):
I like how this consolidates the code and makes the test environment more streamlined. Thanks for improving the docs as well!

(I've not actually tested it, but the docs/code look good to me.)

kshtsk commented on .wasser/config.yaml:
```diff
@@ -55,10 +55,9 @@ routines:
-        cd setups/foo
-        vagrant up --no-destroy-on-error --no-parallel 2>&1
+        cd aquarium
+        ./tools/aqrdev create foo
```
add `2>&1` at the end to see why it is failing...

jecluis replied:
ack

jecluis commented:
@kshtsk mind helping me out here? This is failing because of a `cd aquarium`, it seems, and I don't exactly know why. I tried mimicking the previous behavior, but maybe it was wrong?

kshtsk commented on .wasser/config.yaml:
```diff
@@ -55,10 +55,9 @@ routines:
-        vagrant up --no-destroy-on-error --no-parallel 2>&1
+        cd aquarium
+        ./tools/aqrdev create foo
+        ./tools/aqrdev start foo --conservative
```
and here 2>&1

kshtsk commented on tests/vagrant/templates/Vagrantfile.tmpl:
```diff
@@ -2,15 +2,15 @@
     config.vm.define :"node01", primary: true do |node|
         node.vm.hostname = "node01"
-        node.vm.network "forwarded_port", guest: 1337, host: 1337, host_ip: "*"
+        node.vm.network "forwarded_port", guest: 1337, host: 1337
```
why do you remove this?

jecluis replied:
I didn't. must have been the move or something. I'll check it out.

kshtsk commented on tests/vagrant/templates/Vagrantfile.tmpl:
```diff
@@ -24,7 +24,7 @@ Vagrant.configure("2") do |config|
     config.vm.define :"node02" do |node|
         node.vm.hostname = "node02"
-        node.vm.network "forwarded_port", guest: 1337, host: 1338, host_ip: "*"
+        node.vm.network "forwarded_port", guest: 1337, host: 1338
```
and this?

jecluis replied:
isn't that something you changed recently? I may still need to rebase this on top of main.

kshtsk replied:
yes... so you need to rebase then

kshtsk commented:
```
+ cd aquarium
+ ./tools/aqrdev create foo 2>&1
+ ./tools/aqrdev start foo --conservative 2>&1
>>> bash: ./tools/aqrdev: python3: bad interpreter: No such file or directory
>>> bash: ./tools/aqrdev: python3: bad interpreter: No such file or directory
ERROR:Failed to create node
```

kshtsk commented on tools/aqrdev:
```diff
@@ -0,0 +1,805 @@
+#!python3
```
this odd

jecluis replied:
wth

kshtsk commented:
that's better
but 'click' dependency is not resolved

jecluis commented:
> that's better
> but 'click' dependency is not resolved

this should have been met by uvicorn, which depends on click, but alright, I'll add that to the requirements.txt :)

jecluis commented:
> > that's better
> > but 'click' dependency is not resolved
> 
> this should have been met by uvicorn, which depends on click, but alright, I'll add that to the requirements.txt :)

nope, it's already there.

kshtsk commented on .wasser/config.yaml:
```diff
@@ -55,10 +55,9 @@ routines:
-        cd setups/foo
-        vagrant up --no-destroy-on-error --no-parallel 2>&1
+        cd aquarium
+        ./tools/aqrdev create foo 2>&1
```
Suggestion why not make 'aqrdev' to work without a need to change dir to  aquarium?

jecluis replied:
at this point it requires to have the git repository somewhere as a parent, otherwise it will break. That's something I'm not entirely sure how to fix without having the user specifying the root of the repository.

kshtsk replied:
I usually calculate the root of repo from the __file__ path as a relation.

jecluis replied:
yeah, that is one of the options, but I'm not so sure how that would play out if the file is moved in the future. Something to kick down the road and think about later on.

kshtsk replied:
well, when you move the file around, you tweak the relational path add or remove `..` :-D

jecluis commented:
Oh, I see. The tests are not running out of a virtualenv. I'm unsure if we want to have this stuff being installed in the user's machine just to do development work -- after all, that's what the virtualenv is there for. Any thoughts?

kshtsk commented:
What about use name `acrylic` instead of `aqrdev` and move it from tools to a separate directory?

jecluis commented:
> What about use name `acrylic` instead of `aqrdev` and move it from tools to a separate directory?

I'm not opposed to renaming it and putting it in a different directory, but I'm not seeing how that will fix the dependency/virtualenv issue?

kshtsk commented:
> > What about use name `acrylic` instead of `aqrdev` and move it from tools to a separate directory?
> 
> I'm not opposed to renaming it and putting it in a different directory, but I'm not seeing how that will fix the dependency/virtualenv issue?

it is not related suggestion to the dependency issue

jecluis commented:
> > > What about use name `acrylic` instead of `aqrdev` and move it from tools to a separate directory?
> > 
> > 
> > I'm not opposed to renaming it and putting it in a different directory, but I'm not seeing how that will fix the dependency/virtualenv issue?
> 
> it is not related suggestion to the dependency issue

Alright, Sorry ;)

kshtsk commented:
> Oh, I see. The tests are not running out of a virtualenv. I'm unsure if we want to have this stuff being installed in the user's machine just to do development work -- after all, that's what the virtualenv is there for. Any thoughts?

drop using `click` and standard `argparse`?

jecluis commented:
> > Oh, I see. The tests are not running out of a virtualenv. I'm unsure if we want to have this stuff being installed in the user's machine just to do development work -- after all, that's what the virtualenv is there for. Any thoughts?
> 
> drop using `click` and use the standard `argparse`?

This does not solve the wider problem of not being able to add new dependencies to this script or any other dev script the CI infra might use.

jecluis commented:
> but if you're asking me if we need another dependency resolution for the new tool, then I don't know what you had in mind dropping the setup-dev.sh script.

I didn't drop the `setup-dev.sh` script. I dropped the `setup.sh` script and replaced it with this.

> So we either need to add to FZTH doc how to use this tool. Btw, when you move your python tool to a separate directory there can be a setup config where you can provide all dependencies for it and you just need to install it to virtual environment as a develop target, the same as I did for wasser, but it is probably not what you want

What about, instead of doing this kind of foo, we activate the venv to run this command in the wasser yaml, and deactivate afterwards? We should have a venv as setup earlier by the `setup-dev.sh` script, right?

kshtsk commented on .wasser/config.yaml:
```diff
@@ -55,10 +55,11 @@ routines:
+        source venv/bin/activate
+        ./tools/aqrdev create foo 2>&1
+        ./tools/aqrdev start foo --conservative 2>&1
+        deactivate
```
You do not need to deactivate it since, this channel is dropped, every new command needs own env

jecluis replied:
alright, will adjust.

kshtsk replied:
yeah... for illustration it looks okay to leave it here to correspond steps in FZTH

jecluis replied:
Does this mean this command does not have a virtualenv without running `./setup-dev.sh` first?

kshtsk commented:
> What about, instead of doing this kind of foo, we activate the venv to run this command in the wasser yaml, and deactivate afterwards? We should have a venv as setup earlier by the `setup-dev.sh` script, right?

that makes sense to me

kshtsk commented:
but still, we need to add click to the dev venv

kshtsk commented:
or just python3-click to system?

jecluis commented:
> but still, we need to add click to the dev venv

it's already there: https://github.com/aquarist-labs/aquarium/blob/main/src/requirements.txt#L2

kshtsk commented:
Ah... I see, because you put "#!/bin/python3" or something it always takes system python3 instead of venv. I suggest you to replace it with `#!env python3`

kshtsk commented:
sorry, I meant `#!/usr/bin/env python3`

kshtsk commented:
now it is not clear how to ssh to a box and run commands in it?

jecluis commented:
> now it is not clear how to ssh to a box and run commands in it?

no, it was a mistake. the location of the deployments changed and I forgot to change that in the wasser config.

jecluis commented:
Hurray!

kshtsk commented:
excellent

tserong reviewed (approved):
As Lars said, I like that this simplifies running Vagrant. With the caveat that I haven't actually run it myself, but that the tests all pass, LGTM :-)
